File: src/utils/read_func.R
WORKDIR <- fs::path(Sys.getenv("HOME"), "workspace", "svgbit-test")
reticulate::use_virtualenv(fs::path(WORKDIR, "venv"))

read_mba <- function(sample) {
    # count_df: genes * spots
    gsm <- mba_files[[sample]]
    count_path <- fs::path(
        WORKDIR,
        "data",
        "2020_SciAdv_MouseBrainAtlus",
        "COUNT",
        paste0(gsm, "_expr_raw_counts_table_", sample, ".tsv.gz")
    )
    count_df <- read.csv(count_path, sep = "\t", row.names = 1)
    count_df <- t(count_df)

    x_coor <- c()
    y_coor <- c()
    for (spot_name in colnames(count_df)) {
        spot_name <- strsplit(spot_name, "x")[[1]]
        x_coor <- c(x_coor, as.numeric(strsplit(spot_name[1], "_")[[1]][2]))
        y_coor <- c(y_coor, as.numeric(spot_name[2]))
    }
    coor_df <- data.frame("X" = x_coor, "Y" = y_coor)
    array_df <- data.frame("array_col" = x_coor, "array_row" = y_coor)
    rownames(coor_df) <- colnames(count_df)

    returns <- list(
        "sample" = sample,
        "source" = "mba",
        "count_df" = count_df,
        "coor_df" = coor_df,
        "array_df" = array_df,
        "label" = read_mba_label(sample)
    )
    return(returns)
}

read_mba_label <- function(sample) {
    meta_path <- fs::path(
        WORKDIR,
        "data",
        "2020_SciAdv_MouseBrainAtlus",
        "META",
        paste0(mba_files[[sample]], "_meta_table_", sample, ".tsv.gz")
    )

    meta_df <- read.table(
        meta_path, header = TRUE, sep = "\t", row.names = 1, check.names = FALSE
    )

    returns <- meta_df$ABA_parent
    names(returns) <- rownames(returns)
    return(returns)
}

list_mba_files <- function() {
    files <- fs::dir_ls(
        fs::path(WORKDIR, "data", "2020_SciAdv_MouseBrainAtlus", "COUNT")
    )
    result_dict <- list()

    for (file in files) {
        file_name <- fs::path_file(file)

        if (startsWith(file_name, "GSM")) {
            match <- regexec("^(GSM\\d+)_.*_(\\d{2}A)\\.tsv\\.gz$", file_name)
            match_result <- regmatches(file_name, match)

            if (length(match_result[[1]]) > 0) {
                gsm <- match_result[[1]][2]
                suffix <- match_result[[1]][3]
                result_dict[[suffix]] <- gsm
            }
        }
    }
    return(result_dict)
}
mba_files <- list_mba_files()

read_stereo <- function(sample) {
    # count_df: genes * spots
    read_path <- fs::path(
        WORKDIR,
        "data",
        "2022_Cell_Stereo-seq",
        "STDS0000058",
        "stomics",
        paste0(sample, ".h5ad")
    )
    adata <- anndata::read_h5ad(read_path)
    count_df <- t(as.data.frame(as.matrix(adata$layers["count"])))
    logcount_df <- t(as.data.frame(as.matrix(adata$X)))
    coor_df <- as.data.frame(adata$obsm$spatial)
    colnames(coor_df) <- c("X", "Y")
    rownames(coor_df) <- colnames(count_df)
    array_df <- coor_df
    colnames(array_df) <- c("array_col", "array_row")

    returns <- list(
        "sample" = sample,
        "source" = "stereo",
        "count_df" = count_df,
        "logcount_df" = logcount_df,
        "coor_df" = coor_df,
        "array_df" = array_df
    )
    return(returns)
}

read_dlpfc <- function(sample) {
    # count_df: genes * spots
    count_path <- fs::path(
        WORKDIR, "data", "spatialLIBD", paste0(sample, "-counts.csv")
    )
    logcount_path <- fs::path(
        WORKDIR, "data", "spatialLIBD", paste0(sample, "-logcounts.csv")
    )
    coor_path <- fs::path(
        WORKDIR, "data", "spatialLIBD", paste0(sample, "-coor.csv")
    )
    coldata_path <- fs::path(
        WORKDIR, "data", "spatialLIBD", paste0(sample, "-coldata.csv")
    )

    count_df <- read.csv(count_path, row.names = 1, check.names = FALSE)
    logcount_df <- read.csv(logcount_path, row.names = 1, check.names = FALSE)
    coor_df <- read.csv(coor_path, row.names = 1)
    colnames(coor_df) <- c("X", "Y")

    coldata <- read.csv(coldata_path, row.names = 1)
    coldata <- coldata[complete.cases(coldata[, "spatialLIBD"]),]
    count_df <- count_df[, rownames(coldata)]
    logcount_df <- logcount_df[, rownames(coldata)]
    coor_df <- coor_df[rownames(coldata), ]
    array_df <- coldata[, c("array_row", "array_col")]

    returns <- list(
        "sample" = sample,
        "source" = "dlpfc",
        "count_df" = count_df,
        "logcount_df" = logcount_df,
        "coor_df" = coor_df,
        "array_df" = array_df
    )
    return(returns)
}

read_iimpact <- function(sample) {
    # count_df: genes * spots
    sample_dir <- paste0(gsub("_", " ", sample), " FFPE data")
    count_path <- fs::path(
        WORKDIR,
        "data",
        "2024_GB_iIMPACT",
        sample_dir,
        paste0("10x_", sample, "_ffpe_count.csv")
    )
    count_df <- read.csv(count_path, row.names = 1, check.names = FALSE)
    count_df <- count_df[, colSums(count_df) > 0]
    coor_path <- fs::path(
        WORKDIR,
        "data",
        "2024_GB_iIMPACT",
        sample_dir,
        paste0("10x_", sample, "_ffpe_loc.csv")
    )
    coor_df <- read.csv(coor_path, row.names = 1, check.names = FALSE)
    coor_df$index <- paste(coor_df$x, coor_df$y, sep = "_")
    rownames(coor_df) <- coor_df$index
    rownames(count_df) <- coor_df$index
    coor_df <- coor_df[, c("x", "y")]
    colnames(coor_df) <- c("X", "Y")

    label <- read_iimpact_label(sample)
    count_df <- count_df[names(label), ]
    coor_df <- coor_df[names(label), ]
    count_df <- t(count_df)

    dist_matrix <- as.matrix(dist(coor_df))
    min_distance <- min(dist_matrix[dist_matrix > 0])
    scaling_factor <- 1 / min_distance
    array_df <- ceiling(coor_df * scaling_factor)
    min_coords <- apply(array_df, 2, min)
    array_df <- as.data.frame(array_df - min_coords + c(1, 1))
    colnames(array_df) <- c("array_col", "array_row")
    rownames(array_df) <- rownames(coor_df)

    returns <- list(
        "sample" = sample,
        "source" = "iimpact",
        "count_df" = count_df,
        "coor_df" = coor_df,
        "array_df" = array_df,
        "label" = label
    )
    return(returns)
}

read_iimpact_label <- function(sample) {
    sample_dir <- paste0(gsub("_", " ", sample), " FFPE data")
    coor_path <- fs::path(
        WORKDIR,
        "data",
        "2024_GB_iIMPACT",
        sample_dir,
        paste0("10x_", sample, "_ffpe_manual_annotation.csv")
    )
    coor_df <- read.csv(coor_path, row.names = 1, check.names = FALSE)
    coor_df$index <- paste(coor_df$x, coor_df$y, sep = "_")
    rownames(coor_df) <- coor_df$index
    returns <- coor_df$annotation
    names(returns) <- coor_df$index
    return(returns)
}

read_self <- function(sample) {
    # count_df: genes * spots
    file_name <- "filtered_feature_bc_matrix"
    count_path <- fs::path(
        WORKDIR, "data", "mouse-brain-full", sample, "outs", file_name, "matrix.mtx.gz"
    )
    features_path <- fs::path(
        WORKDIR, "data", "mouse-brain-full", sample, "outs", file_name, "features.tsv.gz"
    )
    barcodes_path <- fs::path(
        WORKDIR, "data", "mouse-brain-full", sample, "outs", file_name, "barcodes.tsv.gz"
    )
    coor_path <- fs::path(
        WORKDIR, "data", "mouse-brain-full", sample, "outs",
        "spatial", "tissue_positions_list.csv"
    )

    count_df <- as.matrix(Matrix::readMM(gzfile(count_path)))
    features <- read.table(gzfile(features_path), header = FALSE, sep = "\t")$V2
    barcodes <- read.table(gzfile(barcodes_path), header = FALSE, sep = "\t")$V1
    rownames(count_df) <- make.unique(features)
    colnames(count_df) <- barcodes

    coor_df <- read.csv(coor_path, row.names = 1, header = FALSE)
    coor_df <- coor_df[colnames(count_df), ]
    array_df <- coor_df[, c("V3", "V2")]
    coor_df <- coor_df[, c("V5", "V4")]
    colnames(coor_df) <- c("X", "Y")
    colnames(array_df) <- c("array_col", "array_row")

    returns <- list(
        "sample" = sample,
        "source" = "self",
        "count_df" = count_df,
        "coor_df" = coor_df,
        "array_df" = array_df
    )
    return(returns)
}

read_self.2 <- function(sample) {
    # count_df: genes * spots
    count_path <- fs::path(
        WORKDIR, "data", "mouse-brain-full", "Data", "scale_df", "raw",
        paste0(sample, "-raw.csv")
    )
    coor_path <- fs::path(
        WORKDIR, "data", "mouse-brain-full", "Data", "coor_df",
        fs::path(sample, "-coor.csv")
    )

    count_df <- read.csv(count_path, row.names = 1, check.names = FALSE)
    coor_df <- read.csv(coor_path, row.names = 1)
    colnames(coor_df) <- c("X", "Y")
    coor_df <- coor_df[colnames(count_df), ]

    returns <- list(
        "sample" = sample,
        "source" = "self",
        "count_df" = count_df,
        "coor_df" = coor_df
    )
    return(returns)
}

read_gbm <- function(sample) {
    # count_df: genes * spots
    file_name <- "filtered_feature_bc_matrix"
    if (sample == "21B-603-5") file_name <- "raw_feature_bc_matrix"
    count_path <- fs::path(
        WORKDIR, "data", "gbm", sample, "outs", file_name, "matrix.mtx.gz"
    )
    features_path <- fs::path(
        WORKDIR, "data", "gbm", sample, "outs", file_name, "features.tsv.gz"
    )
    barcodes_path <- fs::path(
        WORKDIR, "data", "gbm", sample, "outs", file_name, "barcodes.tsv.gz"
    )
    coor_path <- fs::path(
        WORKDIR, "data", "gbm", sample, "outs", "spatial", "tissue_positions.csv"
    )

    count_df <- as.matrix(Matrix::readMM(gzfile(count_path)))
    features <- read.table(gzfile(features_path), header = FALSE, sep = "\t")$V2
    barcodes <- read.table(gzfile(barcodes_path), header = FALSE, sep = "\t")$V1
    rownames(count_df) <- make.unique(features)
    colnames(count_df) <- barcodes

    coor_df <- read.csv(coor_path, row.names = 1, header = TRUE)
    coor_df <- coor_df[colnames(count_df), ]
    array_df <- coor_df[, c("array_col", "array_row")]
    coor_df <- coor_df[, c("pxl_col_in_fullres", "pxl_row_in_fullres")]
    colnames(coor_df) <- c("X", "Y")

    returns <- list(
        "sample" = sample,
        "source" = "gbm",
        "count_df" = count_df,
        "coor_df" = coor_df,
        "array_df" = array_df
    )
    return(returns)
}

File: src/utils/read_func.py
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Callable, Optional
from PIL import Image

import pandas as pd
import scanpy as sc

import svgbit as sb

Image.MAX_IMAGE_PIXELS = None
WORKDIR = Path.joinpath(Path.home(), "workspace", "svgbit-test")


@dataclass
class ReadData:
    sample: str
    source: str
    count_df: pd.DataFrame
    coor_df: pd.DataFrame
    array_df: pd.DataFrame = None
    logcount_df: Optional[pd.DataFrame] = None
    label: Optional[pd.Series] = None
    image: Optional[Callable[[str], Image.Image]] = None


def read_mba(sample):
    # count_df: spots * genes
    count_path = Path.joinpath(
        WORKDIR,
        "data",
        "2020_SciAdv_MouseBrainAtlus",
        "COUNT",
        f"{mba_files[sample]}_expr_raw_counts_table_{sample}.tsv.gz",
    )
    count_df = pd.read_csv(
        count_path,
        index_col=0,
        header=0,
        sep="\t",
        compression="gzip",
    ).T
    count_df = count_df[~count_df.index.duplicated(keep="first")].T
    count_df = count_df.loc[:, count_df.sum() > 0]

    coor_path = Path.joinpath(
        WORKDIR,
        "data",
        "2020_SciAdv_MouseBrainAtlus",
        "COORDINATES",
        f"coordinates_{sample}.tsv",
    )
    coor_df = pd.read_csv(coor_path, index_col=0, header=0, sep="\t")
    coor_df.columns = ["X", "Y"]

    returns = ReadData(
        sample=sample,
        source="mba",
        count_df=count_df,
        coor_df=coor_df,
        label=read_mba_label(sample),
        image=read_mba_image,
    )

    return returns


def read_mba_label(sample):
    meta_path = Path.joinpath(
        WORKDIR,
        "data",
        "2020_SciAdv_MouseBrainAtlus",
        "META",
        f"{mba_files[sample]}_meta_table_{sample}.tsv.gz",
    )
    meta_df = pd.read_csv(
        meta_path,
        index_col=0,
        header=0,
        sep="\t",
        compression="gzip",
    )
    returns = meta_df["ABA_parent"]
    returns.name = "label"

    return returns


def read_mba_image(gsm, sample):
    image_path = Path.joinpath(
        WORKDIR, f"data/2020_SciAdv_MouseBrainAtlus/HE/{gsm}_HE_{sample}.jpg")
    return Image.open(image_path)


def list_mba_files():
    folder_path = Path.joinpath(
        WORKDIR,
        "data",
        "2020_SciAdv_MouseBrainAtlus",
        "COUNT",
    )
    result_dict = {}
    for file in folder_path.iterdir():
        if file.is_file() and file.name.startswith("GSM"):
            match = re.match(r"(GSM\d+)_.*_(\d{2}A)\.tsv\.gz", file.name)
            if match:
                gsm = match.group(1)
                suffix = match.group(2)
                result_dict[suffix] = gsm
    return result_dict


mba_files = list_mba_files()


def read_stereo(sample):
    # count_df: spots * genes
    adata = sc.read_h5ad(
        f"data/2022_Cell_Stereo-seq/STDS0000058/stomics/{sample}.h5ad")
    count_df = pd.DataFrame.sparse.from_spmatrix(
        adata.layers["count"],
        index=adata.obs.index,
        columns=adata.var.index,
    ).sparse.to_dense()
    count_df = count_df.loc[:, count_df.sum() > 0]

    coor_df = pd.DataFrame(adata.obsm["spatial"])
    coor_df.index = count_df.index
    coor_df.columns = ["X", "Y"]

    returns = ReadData(
        sample=sample,
        source="stereo",
        count_df=count_df,
        coor_df=coor_df,
        label=read_stereo_label(sample),
    )

    return returns


def read_stereo_label(sample):
    adata = sc.read_h5ad(
        f"data/2022_Cell_Stereo-seq/STDS0000058/stomics/{sample}.h5ad")

    returns = adata.obs["annotation"]
    returns.name = "label"

    return returns


def read_dlpfc(sample):
    # count_df: spots * genes
    count_path = Path.joinpath(
        WORKDIR,
        f"data/spatialLIBD/{sample}-counts.csv",
    )
    count_df = pd.read_csv(count_path, index_col=0, header=0).T
    count_df = count_df.loc[:, count_df.sum() > 0]
    logcount_path = Path.joinpath(
        WORKDIR,
        f"data/spatialLIBD/{sample}-logcounts.csv",
    )
    logcount_df = pd.read_csv(logcount_path, index_col=0, header=0).T
    coor_path = Path.joinpath(
        WORKDIR,
        f"data/spatialLIBD/{sample}-coor.csv",
    )
    coor_df = pd.read_csv(coor_path, index_col=0, header=0)
    coor_df.columns = ["X", "Y"]

    label_series = read_dlpfc_label(sample)
    label_series = label_series.reindex(index=count_df.index)
    na_spots = label_series[label_series.isna()].index
    count_df = count_df.drop(index=na_spots)
    coor_df = coor_df.drop(index=na_spots)
    logcount_df = logcount_df.drop(index=na_spots)

    returns = ReadData(
        sample=sample,
        source="dlpfc",
        count_df=count_df,
        coor_df=coor_df,
        logcount_df=logcount_df,
        label=read_dlpfc_label(sample),
        image=read_dlpfc_image,
    )

    return returns


def read_dlpfc_label(sample):
    coldata_path = Path.joinpath(
        WORKDIR,
        f"data/spatialLIBD/{sample}-coldata.csv",
    )
    coldata_df = pd.read_csv(coldata_path, index_col=0, header=0)

    returns = coldata_df["spatialLIBD"]
    returns.name = "labe"

    return returns


def read_dlpfc_image(sample):
    image_path = Path.joinpath(WORKDIR,
                               f"data/spatialLIBD/{sample}_full_image.tif")
    return Image.open(image_path)


def read_iimpact(sample):
    # count_df: spots * genes
    sample_dir = f"{sample.replace('_', ' ')} FFPE data"
    count_path = Path.joinpath(
        WORKDIR,
        "data",
        "2024_GB_iIMPACT",
        sample_dir,
        f"10x_{sample}_ffpe_count.csv",
    )
    count_df = pd.read_csv(count_path, index_col=0, header=0)
    count_df = count_df.loc[:, count_df.sum() > 0]
    coor_path = Path.joinpath(
        WORKDIR,
        "data",
        "2024_GB_iIMPACT",
        sample_dir,
        f"10x_{sample}_ffpe_loc.csv",
    )
    coor_df = pd.read_csv(coor_path, index_col=0, header=0)
    coor_df.index = [f"{data.x}_{data.y}" for _, data in coor_df.iterrows()]
    coor_df.columns = ["X", "Y"]
    count_df.index = coor_df.index

    label = read_iimpact_label(sample)
    count_df = count_df.reindex(index=label.index)
    coor_df = coor_df.reindex(index=label.index)

    returns = ReadData(
        sample=sample,
        source="iimpact",
        count_df=count_df,
        coor_df=coor_df,
        label=label,
    )

    return returns


def read_iimpact_label(sample):
    sample_dir = f"{sample.replace('_', ' ')} FFPE data"
    coor_path = Path.joinpath(
        WORKDIR,
        "data",
        "2024_GB_iIMPACT",
        sample_dir,
        f"10x_{sample}_ffpe_manual_annotation.csv",
    )
    coor_df = pd.read_csv(coor_path, index_col=0, header=0)
    coor_df.index = [f"{data.x}_{data.y}" for _, data in coor_df.iterrows()]

    returns = coor_df["annotation"]
    returns.name = "label"

    return returns


def read_10x(sample):
    # count_df: spots * genes
    d = sb.load_10X(f"data/10XGenomics/{sample}/outs")
    count_df = d.count_df.sparse.to_dense()
    count_df = count_df.loc[:, count_df.sum() > 0]
    coor_df = d.coor_df
    returns = ReadData(
        sample=sample,
        source="10X",
        count_df=count_df,
        coor_df=coor_df,
    )
    return returns


def read_self(sample):
    path = Path.joinpath(WORKDIR, f"data/mouse-brain-full/{sample}/outs")
    d = sb.load_10X(path, make_sparse=False)
    d = sb.filters.low_variance_filter(d, 0)
    returns = ReadData(
        sample=sample,
        source="self",
        count_df=d.count_df,
        coor_df=d.coordinate_df,
    )
    return returns


def read_gbm(sample):
    path = Path.joinpath(WORKDIR, "data", "gbm", sample, "outs")
    if sample == "21B-603-5":
        file_name = "raw_feature_bc_matrix"
    else:
        file_name = "filtered_feature_bc_matrix"
    d = sb.load_10X(path, file_name=file_name, make_sparse=False)
    d = sb.filters.low_variance_filter(d, 0)
    returns = ReadData(
        sample=sample,
        source="self",
        count_df=d.count_df,
        coor_df=d.coordinate_df,
    )
    return returns


def strip_zero(str_in):
    str_in = str(str_in)
    while 1:
        if str_in[-1] == "0":
            str_in = str_in[:-1]
        else:
            return str_in
